Retry mock server start on failure

twometresteve opened the pull request:
## Goal

Adds logic to retry starting of the mock server, which can occasionally for more than one (not necessarily well understood) reason.

## Design

Technically it's possible for an infinite loop to occur now, but the build step will timeout if that happens.

The move to Ruby 2.5 was needed to allow use of `rescue` directly beneath a `loop do`.

## Tests

I've tested this manually by temporarily inserting an explicit exception based on a rand() number.  I've also broadened the range of Android devices that we run on as standard, as part of a general attempt to improve the testing that we during development (on the basis that the tests do not take very long).

twometresteve commented on bugsnag-maze-runner.gemspec:
```diff
@@ -7,7 +7,7 @@ Gem::Specification.new do |spec|
   spec.authors = ['Steve Kirkland']
   spec.email   = ['steve@bugsnag.com']
-  spec.required_ruby_version = '>= 2.3.0'
+  spec.required_ruby_version = '>= 2.5.0'
```
It's the `rescue` in a `do`/`end` that was added in 2.5, as described in https://blog.bigbinary.com/2017/10/24/ruby-2.5-allows-rescue-inside-do-end-blocks.html.  (Which I only know because Rubocop pointed it out to me).

I think we're likely to add more `rescue`s as we build more resilience into the system, so it would be good to step up to 2.5 to benefit from the more succinct form.  Unless of course there are reasons for keeping to 2.3 specifically?

Cawllec reviewed (commented):
Small concern as discussed over Slack regarding the loop not ending if the server is unable to start.  While it will eventually time out in Buildkite, it would be better not to take up an agent resource for the entire time.

As such we should add an attempt count and then cause the process to error after a certain number of attempts.  This will allow us to manually restart quickly in the case we have a bad Buildkite agent.

twometresteve commented:
> Small concern as discussed over Slack regarding the loop not ending if the server is unable to start. While it will eventually time out in Buildkite, it would be better not to take up an agent resource for the entire time.
> 
> As such we should add an attempt count and then cause the process to error after a certain number of attempts. This will allow us to manually restart quickly in the case we have a bad Buildkite agent.

No problem, I'll add a retry limit and unit test to back it up.

twometresteve commented:
Unit test for `server` now added.

`hooks.rb` has also changed location from `support` to `steps`.  Its original placement in was causing issues with `require`s.  At runtime I either got duplicate definition warnings or unknown names (both because we pass a set of `--require` options to Cucumber). I eventually found that Cucumber loads all files in `support` first and so the simplest thing was to move it into `steps`, which probably makes sense anyway now I think about it.

Also to note that some of the changes here will look familiar, as I forgot to change the target branch for #121 to `master` before merging.

Cawllec reviewed (approved):
LGTM, looks like there're a couple of things missing from the changelog, but I can handle those for release. 👍
